core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java

```diff
@@ -399,4 +410,64 @@ private static boolean isSplitLocal(List<HostAddress> splitAddresses, HostAddres
         }
         return false;
     }
+
+    private static class QueueSizeAdjuster
+    {
+        private final Map<String, Long> splitsWeightPerNode;
+        private final Set<String> previousScheduleFullNodes;
+        private final Map<String, Long> previouslyScheduledWeight;
+        private final long minPendingSplitsWeightPerTask;
+        private final long maxPendingSplitsWeightPerTask;
+        private final long maxSplitsWeightPerNode;
+
+        QueueSizeAdjuster(long minPendingSplitsWeightPerTask, long maxPendingSplitsWeightPerTask, long maxSplitsWeightPerNode)
+        {
+            this.splitsWeightPerNode = new HashMap<>();
+            this.previousScheduleFullNodes = new HashSet<>();
+            this.previouslyScheduledWeight = new HashMap();
+            this.maxPendingSplitsWeightPerTask = maxPendingSplitsWeightPerTask;
+            this.minPendingSplitsWeightPerTask = minPendingSplitsWeightPerTask;
+            this.maxSplitsWeightPerNode = maxSplitsWeightPerNode;
+        }
+
+        public void update(List<RemoteTask> existingTasks, NodeAssignmentStats nodeAssignmentStats)
+        {
+            if (maxPendingSplitsWeightPerTask == minPendingSplitsWeightPerTask) {
+                return;
+            }
+            for (RemoteTask task : existingTasks) {
+                String nodeId = task.getNodeId();
+                long adjustedMaxSplitsWeightPerTask = getAdjustedMaxPendingSplitsWeightPerTask(nodeId);
+
+                if (adjustedMaxSplitsWeightPerTask < maxPendingSplitsWeightPerTask
+                        && previousScheduleFullNodes.contains(nodeId)
+                        && nodeAssignmentStats.getQueuedSplitsWeightForStage(nodeId) == 0) {
+                    adjustedMaxSplitsWeightPerTask = Math.min(maxPendingSplitsWeightPerTask, adjustedMaxSplitsWeightPerTask * 2);
+                }
+                else if (adjustedMaxSplitsWeightPerTask > minPendingSplitsWeightPerTask
+                        && !previousScheduleFullNodes.contains(nodeId)
+                        && (previouslyScheduledWeight.getOrDefault(nodeId, 0L) - nodeAssignmentStats.getQueuedSplitsWeightForStage(nodeId)) < (getMaxSplitsWeightPerNode(nodeId) / 2)) {
```

Comment:
this still has a risk that if you call two `schedule` immediatelly after each other then `previouslyScheduledWeight.getOrDefault(nodeId, 0L) - nodeAssignmentStats.getQueuedSplitsWeightForStage(nodeId))` 
will be 0 and this condition would trigger. This is false positive because no splits were really consumed,
but they will be consumed in batch at some point.

Reply:
Risk is real, but in worst case we will end up with minimal pending splits per task as it is now without any adjustment. During scheduling we don't have any metric which has minimal lag and is directly connected to how well node is processing splits.

Reply:
> Risk is real, but in worst case we will end up with minimal pending splits per task as it is now without any adjustment.

That mitigates benefits of this PR